### PR TITLE
[WIP] Correctly handle Ember errors

### PR DESCRIPTION
**Pull request by lolmaus**
For discussion, see #33.

### CODE REVIEW

File: app/instance-initializers/configure-trackjs.js [hunk 1]
```diff
@@ -1,34 +1,40 @@
...
-    } else {
-      trackJs.track(err);
-      Ember.Logger.assert(false, err);
+  Ember.onerror = function (error) {
+    if (error.name === 'TransitionAborted') { return; }
+
+    if (error instanceof Error) {
+      trackJs.track(error);
+      Ember.Logger.error(error.stack);
+      return;
```
Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: Is this explicit return actually required?

Comment (lolmaus) [inline app/instance-initializers/configure-trackjs.js]: It is an alternative to `} else {` that lets you avoid extra nesting.

Wait, don't merge this! I'm still working on it.

Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: :+1: If you could add some tests when all said and done that'd be great. I'd hate to see a regression on such gnarly behaviour.

**Comment by lolmaus** [comment]
So I had to deobfuscate TrackJS in order to figure out what's going on.

Turns out TrackJS uses `.toString()` on the error object before passing it to the backend. For subclasses of `Error` this produces a readable error with a stack (probably returns `error.stack`).

But `POJO.toString()` returns `"[object Object]"`! And Ember Router uses [a plain object](https://github.com/tildeio/router.js/blob/50cf4596af4b519dc8c00ad05610afa4a1e7c401/lib/router/transition.js#L314) for transition errors. :/

So in my PR I do this basically:

```js
    if (error instanceof Error) {
      trackJs.track(error);
      Ember.Logger.error(error.stack);
      return;
    }

    trackJs.track(JSON.stringify(error, null, 2));
    Ember.Logger.error(error);
```

But `JSON.stringify()` fails miserably on circular references, and Ember is rich on objects with circular references (e. g. models). So if a user throws an error and passes an object that references a model, he will see "Object with circular references passed instead of an instance of Error" in TrackJS logs. So I had to handle that.

I also removed `Ember.RSVP.on('error', ...)` because Ember.RSVP already [does a great job handling errors](https://github.com/emberjs/ember.js/blob/27862a18b8773e57c52dbc7141b1f92a505f16ff/packages/ember-runtime/lib/ext/rsvp.js#L58-L77). As you can see, it even ignores transition errors for your convenience!

**Comment by lolmaus** [comment]
Sorry, no new tests. It already took way too much time to figure out, can't afford spending a few more hours on this (and I feel testing will require mocking TrackJS).

**Comment by jherdman** [comment]
> Sorry, no new tests. It already took way too much time to figure out, can't afford spending a few more hours on this (and I feel testing will require mocking TrackJS).

TrackJS is [already mocked](https://github.com/jherdman/ember-cli-trackjs/blob/0665ee493890f07d92a32abdb3fed35ff20ad694/tests/dummy/public/fake-trackjs.js).

If you don't have any further time, that's okay. I'll have to reject this PR and build off of your commit though. Is that okay with you? I should have some time later today.

**Comment by lolmaus** [comment]
> I'll have to reject this PR and build off of your commit though.

Uhm, but if I provide tests, then you'll merge?

**Comment by jherdman** [comment]
Pretty much. Here's the checklist:

* There are a few extraneous changes in this commit that I'd like to see removed to focus it (I'll add comments)
* Tests

Again, I don't mind picking up from here if you're out of time. I really really really appreciate all of the effort you've gone through to track down this bug.

File: app/instance-initializers/configure-trackjs.js [hunk 2]
```diff
@@ -1,34 +1,52 @@
 import Ember from 'ember';
 
 export function initialize(application) {
-  let trackJs = application.container.lookup('service:trackjs');
-  let appVersion = application.container.lookup('application:main').get('version');
+  const trackJs = application.container.lookup('service:trackjs');
 
   trackJs.configure({
-    version: appVersion
+    version: application.container.lookup('application:main').get('version')
```
Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: Please revert this change.

File: app/instance-initializers/configure-trackjs.js [hunk 3]
```diff
@@ -1,34 +1,52 @@
...
-  });
+
+    trackJs.track(serializedError);
+    Ember.Logger.error(error);
+  };
 }
 
 export default {
-  name: 'configure-trackjs',
+  name:       'configure-trackjs',
```
Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: Please revert this change.

File: app/instance-initializers/configure-trackjs.js [hunk 4]
```diff
@@ -1,34 +1,52 @@
...
   });
 
-  // http://docs.trackjs.com/Examples/Integrating_with_Ember
-  Ember.onerror = function (err) {
-    if (err.name === 'TransitionError') {
-      Ember.debug('Ignoring TransitionError:', err);
-    } else {
-      trackJs.track(err);
-      Ember.Logger.assert(false, err);
+  Ember.onerror = error => {
```
Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: Is there a reason to use the fat arrow here?

File: app/instance-initializers/configure-trackjs.js [hunk 5]
```diff
@@ -1,34 +1,52 @@
...
+    // If you pass a POJO to TrackJS, you'll see "[object Object]"
+    // in reports instead of object content. :(
```
Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: That's just plain mean :(

File: app/instance-initializers/configure-trackjs.js [hunk 6]
```diff
@@ -1,34 +1,52 @@
...
-      Ember.debug('Ignoring TransitionError:', err);
-    } else {
-      trackJs.track(err);
-      Ember.Logger.assert(false, err);
+    // Otherwise, we have to serialize it manually in order
+    // to see something better than "[object Object]".
+    let serializedError;
+    try {
+      // This will work for both primitives and POJOs:
+      serializedError = JSON.stringify(error, null, 2);
```
Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: Are the `null` and `2` parameters necessary?

**Comment by lolmaus** [comment]
I'm sorry, I gave it an honest effort and it just won't work. :(

First of all, TrackJS won't track errors in test mode. Probably due to [this](27862a18b8773e57c52dbc7141b1f92a505f16ff).

Then, I've got now idea how to throw an error in a route in such a way that it's tracked by TrackJS and not tracked by QUnit. I assume that `assert.throws()` would hide an error from TrackJS but I can't check it because `assert.throws()` is synchronous. I tried `visit('/').catch(() => {})` but it crashes with a strange error "fn is not a function".

Gonna have some ale now. @jherdman, please take it on from here, we're counting on you!

File: app/instance-initializers/configure-trackjs.js (hunk 4 above)
Comment (lolmaus) [inline app/instance-initializers/configure-trackjs.js]: Well, it's shorter, fewer parens, and we don't bind to it, so... Do we need `function` here?

File: app/instance-initializers/configure-trackjs.js (hunk 6 above)
Comment (lolmaus) [inline app/instance-initializers/configure-trackjs.js]: They tell it to produce a multiline JSON with two-space indentation, rather than cluttered into a one-liner. I've tried it and TrackJS UI deals nicely with mutliline messages, removing linebreaks when it shows them as one-liners.

Comment (jherdman) [inline app/instance-initializers/configure-trackjs.js]: Cool!

**Comment by jherdman** [comment]
Dang :( I'll give 'er and see what I can do.

Poke me if you're at an Ember conf some time. Beers on me.

**Comment by lolmaus** [comment]
I'm applying for a visa on Tuesday. If they let me in, I'll be at Boston Ember Meetup on Jan 14, and it'll also be my birthday. :grin:

I owe so much beers, this is the first beer I earned for myself. So proud! :relaxed:
